### PR TITLE
Update rbac changes to support triggers clusterinterceptors

### PR DESCRIPTION
**Pull request by savitaashture**
# Changes
Modified `edit` clusterrole to bind with `clustertriggerbinding` instead of `rolebinding` in order to satisfy triggers new resource `clusterinterceptors`
[rbac example from triggers repo](https://github.com/tektoncd/triggers/blob/main/examples/rbac.yaml#L54)
[Eventlistener Doc](https://github.com/tektoncd/triggers/blob/main/docs/eventlisteners.md#specifying-the-kubernetes-service-account)

**Issue:**
EL pod crashes with below error
```
E0607 13:14:29.078049       1 reflector.go:127] runtime/asm_amd64.s:1371: Failed to watch *v1alpha1.ClusterInterceptor: failed to list *v1alpha1.ClusterInterceptor: clusterinterceptors.triggers.tekton.dev is forbidden: User "system:serviceaccount:default:pipeline" cannot list resource "clusterinterceptors" in API group "triggers.tekton.dev" at the cluster scope
```

/cc @nikhil-thomas @khrm @vdemeester 

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [ ] Includes [tests](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if functionality changed/added)
- [ ] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [ ] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)

_See [the contribution guide](https://github.com/tektoncd/operator/blob/master/CONTRIBUTING.md) for more details._

# Release Notes

```release-note
From triggers 0.13.0 a new CRD clusterinterceptors introduced and which requires cluster level permission.
In order to give permission at cluster level created new clusterrole and clusterrolebinding for clusterinterceptors
```

### CODE REVIEW

**Comment by savitaashture** [comment]
One thing noticed as part of `ClusterTriggerBinding`

As `pipeline` SA created in each namespace because of that `ClusterTriggerBinding` contains all those ns as part of subjects like below
```
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: edit
subjects:
- kind: ServiceAccount
  name: pipeline
  namespace: default
- kind: ServiceAccount
  name: pipeline
  namespace: openshift
- kind: ServiceAccount
  name: pipeline
  namespace: test
- kind: ServiceAccount
  name: pipeline
  namespace: test1

```

**Review by vdemeester — APPROVED** [review]

File: pkg/reconciler/openshift/tektonconfig/rbac.go [hunk 1]
```diff
@@ -399,9 +428,8 @@ func (r *rbac) createRoleBinding(ctx context.Context, sa *corev1.ServiceAccount)
 		Subjects: []rbacv1.Subject{{Kind: rbacv1.ServiceAccountKind, Name: sa.Name, Namespace: sa.Namespace}},
 	}
 
-	_, err = rbacClient.RoleBindings(sa.Namespace).Create(ctx, rb, metav1.CreateOptions{})
-	if err != nil {
-		logger.Error(err, "creation of 'edit' rolebinding failed, in Namespace", sa.GetNamespace())
+	if _, err := rbacClient.ClusterRoleBindings().Create(ctx, crb, metav1.CreateOptions{}); err != nil {
```
Comment (pradeepitm12) [inline pkg/reconciler/openshift/tektonconfig/rbac.go:431]: Should we return an error here?

File: pkg/reconciler/openshift/tektonconfig/rbac.go [hunk 2]
```diff
@@ -356,40 +386,39 @@ func hasSubject(subjects []rbacv1.Subject, x rbacv1.Subject) bool {
...
-	if err != nil {
+	if _, err := rbacClient.ClusterRoles().Get(ctx, "edit", metav1.GetOptions{}); err != nil {
 		logger.Error(err, "getting clusterRole 'edit' failed")
 		return err
 	}
 
-	rb := &rbacv1.RoleBinding{
+	crb := &rbacv1.ClusterRoleBinding{
 		ObjectMeta: metav1.ObjectMeta{
 			Name:            "edit",
```
Comment (nikhil-thomas) [inline pkg/reconciler/openshift/tektonconfig/rbac.go:423]: is there an already existing clusterRoleBinding with name `edit` by default on OpenShift.
If no, let us rename the ClusterroleBinding to something like `openshift-pipelines-edit`

File: pkg/reconciler/openshift/tektonconfig/rbac.go (hunk 1 above)
Comment (savitaashture) [reply pkg/reconciler/openshift/tektonconfig/rbac.go:489]: Yes 
Updated :+1:

**Comment by savitaashture** [comment]
@nikhil-thomas modified code by adding new clusterrole and clusterrolebinding with view permission for `clusterinterceptors`

File: pkg/reconciler/openshift/tektonconfig/rbac.go [hunk 3]
```diff
@@ -399,9 +402,118 @@ func (r *rbac) createRoleBinding(ctx context.Context, sa *corev1.ServiceAccount)
...
+	}
+	return nil
+}
+
+func (r *rbac) ensureClusterRoleBindings(ctx context.Context, sa *corev1.ServiceAccount) error {
+	logger := logging.FromContext(ctx)
+
+	rbacClient := r.kubeClientSet.RbacV1()
+	logger.Info("finding cluster-role clusterinterceptors-clusterrole")
+	if _, err := rbacClient.ClusterRoles().Get(ctx, "clusterinterceptors-clusterrole", metav1.GetOptions{}); errors.IsNotFound(err) {
```
Comment (nikhil-thomas) [inline pkg/reconciler/openshift/tektonconfig/rbac.go:417]: let us add `openshift-pipelines-` prefix to this name and 
we can skip the `clusterrole` suffix, as it is redundant.

File: pkg/reconciler/openshift/tektonconfig/rbac.go [hunk 4]
```diff
@@ -399,9 +402,118 @@ func (r *rbac) createRoleBinding(ctx context.Context, sa *corev1.ServiceAccount)
...
+		return r.createClusterRole(ctx)
+	}
+
+	logger.Info("finding cluster-role-binding clusterinterceptors-clusterrolebinding")
+
+	viewCRB, err := rbacClient.ClusterRoleBindings().Get(ctx, "clusterinterceptors-clusterrolebinding", metav1.GetOptions{})
```
Comment (nikhil-thomas) [inline pkg/reconciler/openshift/tektonconfig/rbac.go:423]: let us add openshift-pipelines- prefix to this name and
we can skip the clusterrole suffix, as it is redundant.

File: pkg/reconciler/openshift/tektonconfig/rbac.go [hunk 5]
```diff
@@ -399,9 +402,118 @@ func (r *rbac) createRoleBinding(ctx context.Context, sa *corev1.ServiceAccount)
...
+
+	hasSubject := hasSubject(rb.Subjects, subject)
+	if !hasSubject {
+		rb.Subjects = append(rb.Subjects, subject)
+	}
+
+	ownerRef := rb.GetOwnerReferences()
+	if len(ownerRef) == 0 {
+		rb.SetOwnerReferences([]metav1.OwnerReference{r.ownerRef})
+	}
```
Comment (nikhil-thomas) [inline pkg/reconciler/openshift/tektonconfig/rbac.go:450]: hi, if this is not `== 0` ideally we should, append our owner reference to existing ones and then set the updated list.
or we should reset it and set ours, as this is a clusterrole created and managed by us.

I think we should not skip setting ownerRefernce under any circumstances.

i don't mind merging this pr as it is.
However, could you create an issue, mentioning this.

1. if ownerReference already exists, check whether one of the ownerReference (kind and name) is ours
2. else append our owner reference.

File: pkg/reconciler/openshift/tektonconfig/rbac.go [hunk 6]
```diff
@@ -399,9 +402,118 @@ func (r *rbac) createRoleBinding(ctx context.Context, sa *corev1.ServiceAccount)
...
+}
+
+func (r *rbac) createClusterRoleBinding(ctx context.Context, sa *corev1.ServiceAccount) error {
+	logger := logging.FromContext(ctx)
+
+	logger.Info("create new clusterrolebinding clusterinterceptors-clusterrolebinding")
+	rbacClient := r.kubeClientSet.RbacV1()
+
+	logger.Info("finding clusterrole clusterinterceptors-clusterrole")
+	if _, err := rbacClient.ClusterRoles().Get(ctx, "clusterinterceptors-clusterrole", metav1.GetOptions{}); err != nil {
```
Comment (nikhil-thomas) [inline pkg/reconciler/openshift/tektonconfig/rbac.go:475]: could you move this name (with changes) into a variable/constant.

File: pkg/reconciler/openshift/tektonconfig/rbac.go [hunk 7]
```diff
@@ -399,9 +402,118 @@ func (r *rbac) createRoleBinding(ctx context.Context, sa *corev1.ServiceAccount)
...
+
+func (r *rbac) createClusterRole(ctx context.Context) error {
+	logger := logging.FromContext(ctx)
+
+	logger.Info("create new clusterrole clusterinterceptors-clusterrole")
+	rbacClient := r.kubeClientSet.RbacV1()
+
+	cr := &rbacv1.ClusterRole{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:            "clusterinterceptors-clusterrole",
```
Comment (nikhil-thomas) [inline pkg/reconciler/openshift/tektonconfig/rbac.go:504]: let us add openshift-pipelines- prefix to this name and
we can skip the clusterrole suffix, as it is redundant.
In addition, let us keep this name in a constant.

File: pkg/reconciler/openshift/tektonconfig/rbac.go (hunk 3 above)
Comment (savitaashture) [reply pkg/reconciler/openshift/tektonconfig/rbac.go]: changed to `openshift-pipelines-clusterinterceptors`

File: pkg/reconciler/openshift/tektonconfig/rbac.go (hunk 5 above)
Comment (savitaashture) [reply pkg/reconciler/openshift/tektonconfig/rbac.go:451]: @nikhil-thomas created issue https://github.com/tektoncd/operator/issues/329 to track this

Also i have cross verified that `ownerReference` added properly

**Comment by nikhil-thomas** [comment]
@savitaashture could you add a short release note.
This is something we should explicity mention in our docs.

**Comment by savitaashture** [comment]
> @savitaashture could you add a short release note.
> This is something we should explicity mention in our docs.

Updated release notes
Thank you

**Pull request merged**